docs/root/intro/arch_overview/advanced/dynamic_modules.rst

```diff
@@ -0,0 +1,69 @@
+.. _arch_overview_dynamic_modules:
+
+Dynamic Modules
+===============
+
+.. attention::
+
+   The dynamic modules feature is an experimental and is currently under active development.
+   Capabilities will be expanded over time and it still lacks some features that are available in other extension mechanisms.
+   We are looking for feedback from the community to improve the feature.
+
+Envoy has support for loading shared libraries at runtime to extend its functionality. We call the shared libraries that
+are loadable by Envoy as "dynamic modules." More specifically, dynamic modules are shared libraries that implement the
+:repo:`ABI <source/extensions/dynamic_modules/abi.h>` written in a pure C header file. The ABI defines a set of functions
+that the dynamic module must implement to be loaded by Envoy. Also, it specifies the functions implemented by Envoy
+that the dynamic module can call to interact with Envoy.
+
+Implementing the ABI from scratch requires an extensive understanding of the Envoy internals. For users, we provide an
+official SDK that abstracts these details and provides a high-level API to implement dynamic modules. The SDK is currently
+available in Rust. In theory, any language that can produce a shared library can be used to implement dynamic modules.
+Future development may include support for other languages.
+
+Currently, the dynamic modules are only supported at the following extension points:
+
+* As an :ref:`HTTP filter  <envoy_v3_api_msg_extensions.filters.http.dynamic_modules.v3.DynamicModuleFilter>`
+
+There are a few design goals for the dynamic modules:
+
+1. **Performance**: The dynamic modules should have minimal overhead compared to the built-in C++ extensions. For example, the dynamic modules are able to access HTTP headers as well as body without copying them unlike any other extension mechanisms.
+2. **Ease of Use**: The SDK should provide a high-level API that abstracts the details of the Envoy internals.
+3. **Flexibility**: The dynamic modules should be able to implement any functionality that can be implemented by the built-in C++ extensions without performance penalty. This is work in progress and many features are not yet available.
+
+Compatibility
+--------------------------
+
+Since the dynamic modules are loaded at runtime, it is important to ensure that the dynamic module is compatible with the
+Envoy binary that loads it.
+
+The dynamic modules have stricker compatibility requirements than the other Envoy extension mechanisms, such as Lua, Wasm or External Processor.
+This is partly because the ABI is tightly coupled with the Envoy internals that makes it a bit difficult to make a stable ABI. Even though
```

Comment:
```suggestion
Stabilizing the ABI is challenging due to the way the ABI needs to be tightly coupled to Envoy's internals. Even though
```

this sentence is hard to understand exactly so im struggling how to reword it - not sure if wording is correct